libmobilecoin/src/transaction.rs

```diff
@@ -1139,6 +1192,352 @@ pub extern "C" fn mc_memo_sender_with_payment_request_memo_get_payment_request_i
+    ffi_boundary_with_error(out_error, || {
+        let memo_data = <[u8; 64]>::try_from_ffi(&gift_code_funding_memo_data)
+            .expect("gift_code_funding_memo_data invalid length");
+        let memo = GiftCodeFundingMemo::from(&memo_data);
+        let key = RistrettoPublic::try_from_ffi(&tx_out_public_key)?;
+
+        let matches = memo.public_key_matches(&key);
+        *out_valid.into_mut() = matches;
+
+        Ok(())
+    })
+}
+
+/// # Preconditions
+///
+/// * `gift_code_funding_memo_data` - must be 64 bytes
+#[no_mangle]
+pub extern "C" fn mc_memo_get_gift_code_funding_note(
+    gift_code_funding_memo_data: FfiRefPtr<McBuffer>,
+) -> FfiOptOwnedStr {
+    ffi_boundary(|| {
+        let memo_data = <[u8; 64]>::try_from_ffi(&gift_code_funding_memo_data)
+            .expect("gift_code_funding_memo_data invalid length");
+        let memo = GiftCodeFundingMemo::from(&memo_data);
+        let note = memo
+            .funding_note()
+            .expect("Could not get gift code funding note");
+
+        FfiOwnedStr::ffi_try_from(note)
+            .expect("Gift code funding note could not be converted a C string")
+    })
+}
+
+/// # Preconditions
+///
+/// * `gift_code_funding_memo_data` - must be 64 bytes
+///
+/// # Errors
+///
+/// * `LibMcError::InvalidInput`
+#[no_mangle]
+pub extern "C" fn mc_memo_get_gift_code_funding_fee(
+    gift_code_funding_memo_data: FfiRefPtr<McBuffer>,
+    out_fee: FfiMutPtr<u64>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let memo_data = <[u8; 64]>::try_from_ffi(&gift_code_funding_memo_data)
+            .expect("gift_code_funding_memo_data has invalid length");
+        let memo = GiftCodeFundingMemo::from(&memo_data);
+        *out_fee.into_mut() = memo.get_fee();
+
+        Ok(())
+    })
+}
+
+/* ==== GiftCodeSenderMemo ==== */
+
+/// # Preconditions
+///
+/// * `fee` - must be an integer less than or equal to 2^56
+/// * `gift_code_sender_note` - must be a null-terminated C string containing up
+/// to 58 valid UTF-8 bytes. The actual note stored on chain is up to 57 null
+/// terminated UTF-8 bytes unless the note is exactly 57 utf-8 bytes long,
+/// in which case, no null bytes are stored. If the C string passed here is
+/// exactly 58 bytes, the last byte MUST be null and that byte will be
+/// removed prior to storage on chain.
+/// * `out_memo_data` - length must be >= 64.
+///
+/// # Errors
+///
+/// * `LibMcError::InvalidInput`
+#[no_mangle]
+pub extern "C" fn mc_memo_gift_code_sender_memo_create(
+    fee: u64,
+    gift_code_sender_note: FfiStr,
+    out_memo_data: FfiMutPtr<McMutableBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let note = <&str>::try_from_ffi(gift_code_sender_note)
+            .expect("Failed to decode gift code sender note");
+        let memo = GiftCodeSenderMemo::new(fee, note).unwrap();
```

Comment:
```suggestion
        let memo = GiftCodeSenderMemo::new(fee, note)
            .expect("Gift code sender note was more than 57 bytes long");

```

Reply:
Same as above, there are multiple failure vectors here that'll get logged by the ffi_boundary